### ISSUE TITLE
Unable to locate element on Real Device post switching to webview of a hybrid app.

### ISSUE THREAD

**Issue by Rahul1970** [appium/appium#16545]
## The problem
Unable to locate element on Real Device post switching to webview of a hybrid app. 

## Environment

* Appium version : 1.22.2
* Desktop OS/version used to run Appium: Windows 10 Android Version 10 
* Node.js version : v16.13.1
* Npm or Yarn package manager: 8.1.2
* Mobile platform/version under test: Android 10
* Real device : Galaxy M20
* Appium CLI or Appium.app|exe:   Appium CLI 

## Details

In my hybrid mobile application i am able to pass through the native views but when landing on webview and able to switch to webview i am not able to locate element.

## Link to Appium logs

<script src="https://gist.github.com/Rahul1970/424fb22182e401f724a308ea8ba65abe.js"></script>

## Code To Reproduce Issue [ Good To Have ]
DesiredCapabilities dc = new DesiredCapabilities();
			dc.setCapability(MobileCapabilityType.DEVICE_NAME, deviceName);//deviceName
			dc.setCapability(MobileCapabilityType.UDID, deviceName );
			//dc.setCapability(MobileCapabilityType.PLATFORM_VERSION,platformVersion);
			dc.setCapability(MobileCapabilityType.PLATFORM_NAME, "Android");
			dc.setCapability(MobileCapabilityType.APP, System.getProperty("user.dir")+"/src/test/resources/apps/"+AppName);
			dc.setCapability("automationName","UiAutomator2");
			dc.setCapability("unicodeKeyboard", true);
			dc.setCapability("resetKeyboard", true);
			dc.setCapability("autoWebview", false);
			//dc.setCapability(MobileCapabilityType.BROWSER_NAME, "Chrome");
			//dc.setCapability("appWaitForLaunch", false);
			dc.setCapability("appPackage", "com.crmnextmobile.crmnextofflineplay");
			dc.setCapability("appActivity", "com.crmnextmobile.crmnextofflineplay.qr.QrScannerActivity");
			driver = new AndroidDriver(service.getUrl(),dc);				

			driver.manage().timeouts().implicitlyWait(Duration.ofSeconds(10));

ReUsableMethods.mobileStartRecording();
        login();
        ReUsableMethods.clickElement(sp.itemMenu," Menu Summary Page ");
        ReUsableMethods.clickElement(lp.leadMenu,"Lead Menu Page");
        ReUsableMethods.clickElement(lp.addLead,"Lead Add icon");
        ReUsableMethods.clickElement(lp.addNewLead,"Add new lead");
        Set<String> contextNames = ((SupportsContextSwitching)driver).getContextHandles(); 
		for (String contextName : contextNames) 
		{
			System.out.println("Available Context"+contextName); 
		}

		System.out.println("Before Context Switch"); ////Switching context to webview
		((SupportsContextSwitching) driver).context("WEBVIEW_com.crmnextmobile.crmnextofflineplay");
		System.out.println("After Context Switch");
		Thread.sleep(10000);
		ReUsableMethods.mobileStartRecording();
		ReUsableMethods.enterText(lp.shortName,Leads[0], "shortname is entered.");
		System.out.println(" Context Switched");
		Thread.sleep(4000);

Hi Team Any help in this direction would be very  helpful.

**Comment by mykola-mokhnach** [comment]
Is the destination element present in the actual page source?

**Comment by Rahul1970** [comment]
Yes it is present in the page source and running fine on emulator. Moreover i have verified its existence on the chrome console as the only unique element present. However not able to inspect on appium inspector but was able to inspect the webview of mobile app  using chrome//urls and chrome//inspect options.
1. Not able to click and inspect element using appium inspector for webview in mobile app.
2. Able to click and inspect using chrome//urls and chrome//inspect options and verified it as the only unique existing element.

**Comment by mykola-mokhnach** [comment]
Please attach the actual xml page source before locating the actual element

**Comment by Rahul1970** [comment]
<input type="text" name="LASTNAME" placeholder="Last Name" class="textbox field__item" maxlength="256" data-autoid="LASTNAME_ctrl" value="">

**Comment by mykola-mokhnach** [comment]
I need the whole xml. It is also important to retrieve it in the moment before the actual element lookup is executed. 

The runtime snapshot of the page might differ from a static one you get in the inspector. In general the element lookup is up to the chromedriver. Appium simply acts as a proxy there. That is why if the element is not present in that xml then there is no way it could be found and there's nothing that could be changed on appium side.

**Comment by KazuCocoa** [comment]
i saw the differences between chromedriver and devtools (chrome's inspector) before. So, potentially interactions via chromedriver (that is working behind appium) differs from devtools based ones in some points

**Comment by Rahul1970** [comment]
Hi @mykola-mokhnach 
As asked i have shared the entire xml page source for your reference. Please suggest what next to be done. Any suggestion in this direction would be helpful.

**Comment by mykola-mokhnach** [comment]
Unfortunately this is not the page source that I need. I require the one generated by chromedriver

**Comment by mykola-mokhnach** [comment]
Also deleted the previous copypaste. Consider providing such info in gists next time as it prevents the actual issue from being read properly

**Comment by Rahul1970** [comment]
Hi mykola,
                 I am new to this please let me know how to get the page source generated by chromedriver.
1. is it by using driver.getpagesource()?
2. is it by using chrome//url and then chrome//inspect in chrome browser?
3. or do i have to share  the appium logs from where the context switch to webview is happening and chromedriver is getting invoked?

**Comment by Rahul1970** [comment]
Hi mykola,
                I am sharing the gist for 
1.    chromedriver get page source: https://gist.github.com/Rahul1970/bf078ac294e534b7749f5e33493bd110.js
2. Appium Logs During Context Switch
 https://gist.github.com/Rahul1970/585e3c08718ec291bf2b0c4452af89b3.js

**Comment by mykola-mokhnach** [comment]
> chromedriver get page source: https://gist.github.com/Rahul1970/bf078ac294e534b7749f5e33493bd110.js

If this is the page source returned by chromedriver then it is expected your locator does not work. I cannot observe such element to be present there.

**Comment by Rahul1970** [comment]
Hi mykola,
                I am sharing the new  gist for chromedriver page source for working code on emulator on successful switch context to webview and able to locate element.
1. https://gist.github.com/Rahul1970/97f8e8fd0a5f1145873a237855ba33c1.js
2.Appium logs for working code on emulator on successful switch context to webview and able to locate element.
   https://gist.github.com/Rahul1970/c503aa9943cf60649a7fed269f5b5797.js
Please suggest how can i get my code to locate element on  real device.

**Comment by mykola-mokhnach** [comment]
Unfortunately we cannot help much on that. Like I said above elements location is fully controlled by chromedriver, which os maintained by Google. Consider reporting your issue to chromedriver's issues tracker if you think this is a bug.

**Comment by Rahul1970** [comment]
Hi Mykola,
                I am sharing with you the getpagesource() response on real device vs emulator for your reference which are completely different.
1. getpagesource() response on Real Device https://gist.github.com/Rahul1970/bf078ac294e534b7749f5e33493bd110.js
2. getpagesource() response on Emulator
https://gist.github.com/Rahul1970/97f8e8fd0a5f1145873a237855ba33c1.js
Please Suggest.

**Comment by jlipps** [comment]
Seems like your app is just in different states in those two situations, or the app behaves differently on real devices vs emulators. This is not something Appium has anything to do with.
